Soc refactor

r0qs opened the pull request:
This PR attempt to refactor the soc API to a more idiomatic and clean API.

The new API exposes the following:
```go
func New(id ID, ch swarm.Chunk) *Soc
func NewSigned(id ID, ch swarm.Chunk, owner, sig []byte) (*Soc, error)

func (s *Soc) Sign(signer crypto.Signer) (swarm.Chunk, error)
func (s *Soc) Chunk() (swarm.Chunk, error)
func (s *Soc) WrappedChunk() swarm.Chunk

func FromChunk(sch swarm.Chunk) (*Soc, error)
func CreateAddress(id ID, owner Owner) (swarm.Address, error) 
func Valid(ch swarm.Chunk) bool
```

acud commented on pkg/soc/soc.go:
```diff
@@ -22,130 +22,131 @@ const (
+func NewSigned(id ID, ch swarm.Chunk, owner, sig []byte) (*Soc, error) {
 	s := New(id, ch)
-	s.signature = sig
 	o, err := NewOwner(owner)
```
I think that `Owner` type is redundant and unnecessary. We could just settle for a simple byte slice and just verify the length in the constructor with
```go
	if len(address) != crypto.AddressSize {
		return nil, fmt.Errorf("soc: invalid owner address %x", address)
	}
```

WDYT?

r0qs replied:
I agree. I'll do the change and remove the Owner type.

acud commented on pkg/soc/soc.go:
```diff
@@ -22,130 +22,131 @@ const (
+		return nil, err
 	}
-	s.signer = signer
 	s.owner = ownerAddress
```
I don't think that we need to mutate this field. The same goes for the signature field. If there is no subsequent usage of the `Soc` instance after signing, then this is probably not necessary.

r0qs replied:
The signature is used in the `Chunk` method to create the soc chunk in the return of the `Sign` function, but I can do it in the `Sign` method instead. 

The `owner` is used to generate the soc's address, but it is currently used only in the tests and in the `Valid` method. Removing the owner would require us to extract the signature from the chunk and make the recovery of the owner address in the `Valid` function and use it there to recreate the chunk address for comparison.

r0qs replied:
For example, I could create a signed soc, and then call `Sign` again with a different signer, and if the sign does not mutate the owner, the newly signed soc will have another owner from the signer and result in a different soc address.

acud commented on pkg/soc/testing/soc.go:
```diff
@@ -0,0 +1,63 @@
+
+// GenerateMockSoc generates a valid mocked soc from given data.
+// If data is nil it generates random data.
+func GenerateMockSoc(data []byte) *MockSoc {
```
Ideally, this should also take a `t *testing.T` as a parameter, then first line of the function call `t.Helper()`.
I would also actually handle the errors, and once the `t.Helper` is called, you can just `t.Fatal(err)`  directly in the error checks

acud reviewed (commented):
This is really great :clap: I have just a few minor suggestions

acud commented on pkg/soc/soc.go:
```diff
@@ -22,130 +21,118 @@ const (
+type ID []byte
 
-// Soc wraps a single soc.
+// Soc wraps a single-owner chunk.
```
actually, `Soc` wraps a content-addressed chunk :)

r0qs replied:
haha of course, sorry. Fixed.

acud reviewed (approved):
Great stuff @r0qs :clap: 
Much needed refactoring. Very nicely done!

acud commented:
![2021-03-16-100506_955x473_scrot](https://user-images.githubusercontent.com/12988138/111283176-206ce080-863f-11eb-93f6-eef2f1d26123.png)

Something came up on DeepSource

r0qs commented:
> ![2021-03-16-100506_955x473_scrot](https://user-images.githubusercontent.com/12988138/111283176-206ce080-863f-11eb-93f6-eef2f1d26123.png)
> 
> Something came up on DeepSource

I added this for convenience in the tests to generate random data for a chunk. I'm not doing any security-related operation with it. So we can ignore this warning.

But looking at my tests again now, I noticed that although I added the possibility of randomly generate the chunk data, I'm not using it anymore. So I'm removing this code since the caller of `GenerateMockSoc` can generate random data if needed.

janos reviewed (approved):
Nice.

I have one small suggestion, and that is to rename `Soc` to `SOC` everywhere as it is an abbreviation. It is also more clear regarding the pronunciation and the usage of indefinite article.

r0qs commented:
> as it is an abbreviation

While renaming, I found this [comment](https://github.com/ethersphere/bee/blob/master/pkg/api/file.go#L304), and I'm wondering what concept this comment is about?
```
// If none match header is set always send the reply as not modified
// TODO: when SOC comes, we need to revisit this concept
```

janos commented:
> what concept this comment is about?

@r0qs I believe that http.StatusNotModified should not be returned in case of SOC chunk, as browser caching may not be desired. But, the concept should be revisited. How to properly do the caching if possible.

acud commented:
I think that the comment might relate to feeds, not SOC chunks, but I might be wrong. I don't see why an SOC should change this logic.
